### PR TITLE
Hunks!

### LINKED ISSUE #127: [TOOL-45] Generate diff metrics from git diff output
GitHub has additions/deletions and changes, while Gitlab doesn't calculate diff stats, so we need to parse diff output to generate our own.

* Write parser for extracting number of lines added/deleted from diff output
* ~~Update extraction process to extract diff metrics~~
* ~~Compare if implemented metric extraction generates same data as GitHub~~

Resources:

[https://gist.github.com/dejan-crocoder/54a317b151764ab7c66c77819ffb7046](https://gist.github.com/dejan-crocoder/54a317b151764ab7c66c77819ffb7046)

<sub>From [SyncLinear.com](https://synclinear.com) | [TOOL-45](https://linear.app/crocoder/issue/TOOL-45/generate-diff-metrics-from-git-diff-output)</sub>

### PR DESCRIPTION
**Pull request by davidabram**
closes #127

### CODE REVIEW

File: packages/integrations/source-control/src/parse-hunks.test.ts [hunk 1]
```diff
@@ -0,0 +1,471 @@
+import { parseHunks } from './parse-hunks';
+import type { Hunk, Change } from './parse-hunks';
+
+const emptyHunk = '';
```
Comment (dejan-crocoder) [inline packages/integrations/source-control/src/parse-hunks.test.ts:4]: hunky monkey

File: packages/integrations/source-control/src/parse-hunks.test.ts [hunk 2]
```diff
@@ -0,0 +1,471 @@
...
++import { SQSClient, SendMessageCommand, SendMessageBatchCommand } from "@aws-sdk/client-sqs";
+ import { z } from "zod";
+ import type { ZodRawShape, ZodAny, ZodObject } from "zod";
+ import { nanoid } from "nanoid";
+ import type { SQSEvent } from "aws-lambda";
+
+-const sqs = new AWS.SQS();
++const sqs = new SQSClient();
+
+ type Content<Shape extends ZodRawShape> = z.infer<ZodObject<Shape, "strip", ZodAny>>
```
Comment (dejan-crocoder) [inline packages/integrations/source-control/src/parse-hunks.test.ts:19]: nice

File: packages/integrations/source-control/src/parse-hunks.test.ts [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+              oldLineNumber: 1,
+              type: 'normal',
+            },
+            {
+              content: 'This is the second line.',
+              lineNumber: 2,
+              type: 'delete',
+            },
+            {
+              content: 'This is the third line.',
```
Comment (dejan-crocoder) [inline packages/integrations/source-control/src/parse-hunks.test.ts:94]: indeed the third line

**Comment by davidabram** [comment]
[I DON'T CARE ABOUT HUNKY BOYS!]( https://youtu.be/Zwrs8PvlKWI?si=3RdwAKJllisO7MmP)

File: packages/integrations/source-control/src/parse-hunks.test.ts [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+ });
+ const extractMergeRequestEventSchema = z.object({
+   mergeRequestIds: z.array(MergeRequestSchema.shape.id),
+-  repositoryId: RepositorySchema.shape.id,
+-  namespaceId: NamespaceSchema.shape.id
+ });
+
+ export type extractMergeRequestsEventMessage = z.infer<typeof extractMergeRequestEventSchema>;`;
+
+describe('parse-hunks:', () => {
```
[the remark this answers is not in the document]
Comment (davidabram) [reply packages/integrations/source-control/src/parse-hunks.test.ts:98]: neither do I. 

I don't know why have I done this.

**Comment by dejan-crocoder** [comment]
Commented on pull request

File: packages/integrations/source-control/src/parse-hunks.ts [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+  const i = array.filter(c => c === 'i').length;
+
+  const c = Math.min(d, i);
+
+  return {
+    d: d - c,
+    i: i - c,
+    c,
+  }
+}
```
Comment (dejan-crocoder) [inline packages/integrations/source-control/src/parse-hunks.ts:56]: pls us lngr nms fr thngs

If d = deletions
And i = insertions
Why is c(hanges?) min of both ?

Comment (davidabram) [reply packages/integrations/source-control/src/parse-hunks.ts:56]: no.

Comment (davidabram) [reply packages/integrations/source-control/src/parse-hunks.ts:56]: These are for future use -> I have just implemented them right now.

These are for analyzing context of an PR -> We will need to find out how long it takes to review a PR.

c -> are number of lines that are deleted and new one inserted in their place. -> my educated guess is that these line are the hardest to read.

That's why dic.

**Review by dejan-crocoder — APPROVED** [review]

**Pull request merged**